Ensure that rc5 doesn't try to use a key longer than 2040 bits

mattcaswell opened the pull request:
The maximum key length for rc5 is 2040 bits so we should not attempt to
use keys longer than this. We truncate them to 2040 bits instead.

Issue found by OSS-Fuzz and Guido Vranken.

mattcaswell commented:
@guidovranken

guidovranken commented:
Confirmed that this resolves the issue.

t-j-h commented:
Silent truncation seems like the *wrong* choice to me. It should fail.

mattcaswell commented:
> Silent truncation seems like the wrong choice to me. It should fail.

That isn't possible to do without an API break because this is a void function. I did consider that, but it doesn't seem appropriate for a stable branch since this is targeting 1.1.1. Silent truncation seemed like the lesser of two evils.

kroeckx commented on doc/man3/EVP_rc5_32_12_16_cbc.pod:
```diff
@@ -43,10 +62,6 @@ These functions return an B<EVP_CIPHER> structure that contains the
-=head1 BUGS
-
-Currently the number of rounds in RC5 can only be set to 8, 12 or 16.
-This is a limitation of the current RC5 code rather than the EVP interface.
```
This still seems to be the case.

mattcaswell replied:
Yes, but I documented the restriction of 8, 12 or 16 in the section on the EVP_CTRL_SET_RC5_ROUNDS ctrl. I didn't see this as a "bug" so the writing it in a bugs section seemed misplaced. The fact that its a limitation of the rc5 code rather than the evp interface is an implementation detail that doesn't seem relevant to the end user.

kroeckx commented:
In 3.0 you could at least change it to return something, that is API compatible. I doubt it's ABI compatible on all our platforms to suddenly return something, so for 1.1 either silent truncation, or just returning would be fine for me.

mattcaswell commented:
> In 3.0 you could at least change it to return something, that is API compatible. 

Yes, ok. That does make sense.

guidovranken commented:
Will this get merged at some point? https://oss-fuzz.com/testcase-detail/5750176758628352 is still active.

guidovranken commented:
Also see https://oss-fuzz.com/testcase-detail/5672061294346240 (RC5 integer overflow) which might be worth fixing.

mattcaswell commented:
Thanks for the reminder. I've reworked and rebased this. There are now 2 commits:

1) The first commit adds a check at the EVP layer for a key that is too long and fails if that occurs, but no changes are made to the underlying `RC5_32_set_key` function.

2) The second commit changes the return type of `RC5_32_set_key` from void to int. The function is checked to ensure that the key is not too long. It returns 1 on success and 0 on error.

The second change should be API compatible but is not ABI compatible. I plan to push both commits to master and only the first commit to the 1.1.1 branch.

This means that, in 1.1.1, using RC5 via EVP will correctly fail if the key is too long. Using the low level RC5 functions directly with a key that is too long will still crash (as it does now).

Ping @kroeckx, @t-j-h.

paulidale commented on doc/man3/EVP_rc5_32_12_16_cbc.pod:
```diff
@@ -33,7 +33,26 @@ EVP_rc5_32_12_16_ofb()
+
+=item EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GET_RC5_ROUNDS, 0, &rounds)
+
+Stores the number of rounds currently configured in B<*rounds> where B<rounds>
```
Shouldn't the second be *rounds too?
Alternatively, say it is a pointer to an int.

levitte commented:
(side note: I'm unsure if changing a void function to a function returning int is ABI incompatible. Such a change only means that old code ignores the return value, right?)

levitte commented:
(ah, except letter downgrades, which is permissible. Never mind)

kroeckx commented:
The return value can actually have an effect on the ABI (not API).

mattcaswell commented:
Pushed with the documentation fixup to master. Also cherry-picked the first commit only to 1.1.1.
